packages/backend/src/invoice/service.ts

```diff
@@ -98,6 +104,39 @@ async function createInvoice(
   }
 }
 
+// Deactivate expired invoices that have some money.
+// Delete expired invoices that have never received money.
+// Returns the id of the processed invoice (if any).
+async function deactivateNextInvoice(
+  deps: ServiceDependencies
+): Promise<string | undefined> {
+  return deps.knex.transaction(async (trx) => {
+    // 30 seconds backwards to allow a prepared (but not yet fulfilled/rejected) packet to finish before being deactivated.
+    const now = new Date(Date.now() - 30_000).toISOString()
+    const invoices = await Invoice.query(trx)
+      .limit(1)
+      // Ensure the invoices cannot be processed concurrently by multiple workers.
+      .forUpdate()
+      // If an invoice is locked, don't wait — just come back for it later.
+      .skipLocked()
+      .where('active', true)
+      .andWhere('expiresAt', '<', now)
+    const invoice = invoices[0]
+    if (!invoice) return
+
+    // Fetch the account with a separate transaction to avoid locking it.
+    const balance = await deps.accountService.getBalance(invoice.accountId)
+    if (balance) {
+      deps.logger.trace({ invoice: invoice.id }, 'deactivating expired invoice')
+      await invoice.$query(trx).patch({ active: false })
+    } else {
+      deps.logger.debug({ invoice: invoice.id }, 'deleting expired invoice')
+      await invoice.$query(trx).delete()
```

Comment:
This will delete the account and invoice if you add the `.onDelete('CASCADE')` from https://github.com/interledger/rafiki/pull/178#discussion_r745036171
```suggestion
      await invoice.$relatedQuery('account', trx).delete()
```